### ISSUE TITLE
OpenShift Support

### ISSUE THREAD

**Issue by jsanchezmartinez** [maksim-paskal/aks-node-termination-handler#75]
Hi,
When running in OpenShift, there are no VirtualMachineScaleSets (only VirtualMachines), and for that reason, the DaemonSet is crashing (attached logs below).
Can we request for OpenShift support?
```
{"file":"github.com/maksim-paskal/aks-node-termination-handler/cmd/main.go:55","func":"main.main","level":"info","msg":"Starting 1.0.13-d8d5a71-1707463489...","time":"2024-03-07T10:48:45Z"}
{"file":"github.com/maksim-paskal/aks-node-termination-handler/pkg/alert/alert.go:29","func":"github.com/maksim-paskal/aks-node-termination-handler/pkg/alert.Init","level":"warning","msg":"not sending Telegram message, no token","time":"2024-03-07T10:48:45Z"}
{"file":"github.com/maksim-paskal/aks-node-termination-handler/pkg/client/client.go:45","func":"github.com/maksim-paskal/aks-node-termination-handler/pkg/client.Init","level":"info","msg":"No kubeconfig file use incluster","time":"2024-03-07T10:48:45Z"}
{"error":"error in getting azure resource name: azure:///subscriptions/dd6b40ef-de5f-4649-95a7-bd2337c71900/resourceGroups/ocp-azure-uat-euw-8npmn-rg/providers/Microsoft.Compute/virtualMachines/master-1: azureProviderID not valid","file":"github.com/maksim-paskal/aks-node-termination-handler/cmd/main.go:86","func":"main.main","level":"fatal","msg":"","time":"2024-03-07T10:48:45Z"}
```

**Comment by maksim-paskal** [comment]
@jsanchezmartinez thanks for opening this issue. By default this tool use AKS (as a Azure service) nodes. All nodes that AKS  creates have `.spec.providerID` that corresponds to this format
https://github.com/maksim-paskal/aks-node-termination-handler/blob/d8d5a71ab0612096d46acab1a6a4c24b454619b5/pkg/api/api.go#L37

I never use OpenShift, but theoretically this tool can works on OpenShift nodes. Can you share some recipe how to build OpenShift cluster on Azure?

I will try to create cluster on Azure, and try to run this tool on this nodes.

**Comment by jsanchezmartinez** [comment]
Hi @maksim-paskal,
Easiest way is using ARO service (https://azure.microsoft.com/en-us/products/openshift/ and https://portal.azure.com/?feature.msaljs=true#view/HubsExtension/BrowseResource/resourceType/Microsoft.RedHatOpenShift%2FOpenShiftClusters).
In OpenShift there are no VirtualMachineScaleSets (only VMs), so the Azure provider ID is a bit different: "^azure:///subscriptions/(.+)/resourceGroups/(.+)/providers/Microsoft.Compute/virtualMachines/(.+)$"

**Comment by maksim-paskal** [comment]
`aks-node-termination-handler` listen all events from `Azure Scheduled Events` and reboot is one of events that Azure sends...

How you plan to use this tool? Are you plan to use Azure Spot?

**Comment by jsanchezmartinez** [comment]
Yes. We are currently running spot instances and we want mainly to drain nodes when eviction events are detected.
Maybe you can test simulating eviction events through Azure API: https://learn.microsoft.com/en-us/rest/api/compute/virtual-machines/simulate-eviction?view=rest-compute-2023-10-02&tabs=HTTP

**Comment by maksim-paskal** [comment]
How can I add to OpenShift cluster spot instances?

**Comment by jsanchezmartinez** [comment]
Basically, you have to pick an existing worker MachineSet, copy/paste it and adapt (https://learn.microsoft.com/en-us/azure/openshift/howto-spot-nodes).
This is the important part to be added:

```
      providerSpec:
        value:
          spotVMOptions: {}
```

**Comment by maksim-paskal** [comment]
Simulation API is not available for OpenShift servers, there are some restrictions on servers resource group . If you have OpenShift cluster with Spots, you can test my change in your cluster.

OpenShift clusters have some [restriction](https://docs.openshift.com/container-platform/4.10/networking/understanding-networking.html) for pods that want to connect to `169.254.169.254`, aks-node-termination-handler needs access to that address for reading events. You must enable this installing chart with `--set hostNetwork=true`

```bash
helm upgrade aks-node-termination-handler \
--install \
--namespace kube-system \
https://github.com/maksim-paskal/aks-node-termination-handler/releases/download/v1.0.13/aks-node-termination-handler-1.1.5.tgz \
--set priorityClassName=system-node-critical \
--set image=paskalmaksim/aks-node-termination-handler:dev \
--set hostNetwork=true
```

If you can test it, this will be awesome - after test I will release that feature.

**Comment by jsanchezmartinez** [comment]
I'll try to test today or next Monday and come back. Thanks :)

**Comment by jsanchezmartinez** [comment]
Seems to be working fine (see attached logs screenshot). Do you need/want anything else to validate?

![image](https://github.com/maksim-paskal/aks-node-termination-handler/assets/137485252/fa7ab951-6b00-49dc-8383-f5577972bf31)

**Comment by maksim-paskal** [comment]
Let's watch. If something wrong happens, please describe that issue. I will release these changes next week (Thursday), if no issues are found.
